File: fn/src/main/java/org/sqlbuilder/fn/joins/FEGroupPattern_FEPattern.java
package org.sqlbuilder.fn.joins;

import org.sqlbuilder.common.Insertable;
import org.sqlbuilder.annotations.RequiresIdFrom;
import org.sqlbuilder.fn.objects.FERealization;
import org.sqlbuilder.fn.objects.ValenceUnit;

import java.util.HashSet;
import java.util.Set;

public class FEGroupPattern_FEPattern extends Triple<FEGroupPattern, FERealization, ValenceUnit> implements Insertable
{
	public static final Set<FEGroupPattern_FEPattern> SET = new HashSet<>();

	// C O N S T R U C T O R

	@SuppressWarnings("UnusedReturnValue")
	public static FEGroupPattern_FEPattern make(final FEGroupPattern groupPattern, final FERealization fer, final ValenceUnit vu)
	{
		var p = new FEGroupPattern_FEPattern(groupPattern, fer, vu);
		SET.add(p);
		return p;
	}

	private FEGroupPattern_FEPattern(final FEGroupPattern groupPattern, final FERealization fer, final ValenceUnit vu)
	{
		super(groupPattern, fer, vu);
	}

	// I N S E R T

	@RequiresIdFrom(type = FEGroupPattern.class)
	@RequiresIdFrom(type = FERealization.class)
	@RequiresIdFrom(type = ValenceUnit.class)
	@Override
	// grouppatternid, ferid, vuid
	public String dataRow()
	{
		return String.format("%s,%s,%s", first.getSqlId(), second.getSqlId(), third.getSqlId());
	}

	@Override
	public String comment()
	{
		return String.format("fegr={%s} fer={%s} vu={%s}", first.comment(), second.comment(), third.comment());
	}

	// T O S T R I N G

	@Override
	public String toString()
	{
		return String.format("[PAT-VU fer=%s pattern=%s, vu=%s]", second, first, first);
	}
}

File: fn/src/main/java/org/sqlbuilder/fn/joins/FEGroupPattern_FEPattern.kt
package org.sqlbuilder.fn.joins

import org.sqlbuilder.annotations.RequiresIdFrom
import org.sqlbuilder.common.Insertable
import org.sqlbuilder.fn.objects.FERealization
import org.sqlbuilder.fn.objects.ValenceUnit
import java.util.*

data class FEGroupPattern_FEPattern(
    val groupPattern: FEGroupPattern,
    val fer: FERealization,
    val vu: ValenceUnit,
) : Insertable {

    // I N S E R T

    @RequiresIdFrom(type = FEGroupPattern::class)
    @RequiresIdFrom(type = FERealization::class)
    @RequiresIdFrom(type = ValenceUnit::class)
    override fun dataRow(): String {
        return "${groupPattern.getSqlId()},${fer.getSqlId()},${vu.getSqlId()}"
    }

    override fun comment(): String {
        return "${groupPattern.comment()} ${fer.comment()} vu={${vu.comment()}}"
    }

    // I D E N T I T Y

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as FEGroupPattern_FEPattern
        return groupPattern == that.groupPattern && fer == that.fer && vu == that.vu
    }

    override fun hashCode(): Int {
        return Objects.hash(groupPattern, fer, vu)
    }

    // T O S T R I N G

    override fun toString(): String {
        return "[PAT-VU fer=$fer pattern=$groupPattern, vu=$vu]"
    }

    companion object {

        @JvmField
        val LIST = ArrayList<FEGroupPattern_FEPattern>()

        @JvmStatic
        fun make(groupPattern: FEGroupPattern, fer: FERealization, vu: ValenceUnit): FEGroupPattern_FEPattern {
            val p = FEGroupPattern_FEPattern(groupPattern, fer, vu)
            LIST.add(p)
            return p
        }
    }
}
